### ISSUE TITLE
All extension network traffic breaks when system proxy settings are changed

### ISSUE THREAD

**Issue by matthewjamesadam** [microsoft/vscode#174279]
Does this issue occur when all extensions are disabled?: No -- this is a bug that affects all extensions

- VS Code Version: 1.75.1
- OS Version: macOS Monterey 12.5.1

Steps to Reproduce:

1. Close VSCode
2. Set a system-level proxy (I do this by running Proxyman)
3. Start VSCode -- note, all VSCode extension traffic correctly goes through the proxy
4. Change the proxy settings (in this case, close Proxyman)
5. All VSCode extension traffic stops working.  It looks like all traffic continues to be routed through the proxy even though the proxy is no longer active.

This appears to affect *all* extensions.  I see this behaviour with our own extension (using `node-fetch`) but also in the GitHub extension.  The error I see in the GitHub extension is `Fetching pull requests failed: request to https://api.github.com/repos/matthewjamesadam/vscode/pulls?per_page=20&page=1 failed, reason: connect ECONNREFUSED 127.0.0.1:9090`.

<img width="450" alt="Screen Shot 2023-02-13 at 11 30 19 AM" src="https://user-images.githubusercontent.com/2133518/218558166-be3d6d23-814a-4ce2-b664-c39597785b97.png">

Note that built-in VSCode network traffic continues to work -- in the extension sidebar I can search the marketplace, for instance.  So this seems to be something particular to extension network traffic.

I am willing to look into fixing this issue if I could get a pointer as to where in the code to look to fix it.  Thanks!

**Comment by isidorn** [comment]
@matthewjamesadam thanks for creating this issue. Just to clarify is this something new that started happening or was it always like this?

fyi @bpasero for potential sandbox work that might affect this as mentioned in standup

**Comment by bpasero** [comment]
Can you try to reproduce with our nightly insider builds? You can give our preview releases a try from: https://code.visualstudio.com/insiders/

**Comment by matthewjamesadam** [comment]
@isidorn I do not think this is a new issue -- I believe I've seen this in the past

@bpasero confirmed, this occurs the same in Insider builds

**Comment by DavyLandman** [comment]
How does it work, this proxy configuration sharing to the extension host process? We've just run into the limitation of nodejs that is using a fixed set of root certificates, instead of respecting the host OS certificate store.

I've been reading the source a bit, and it looks like vscode patches the https & http agent code of node, but I cannot see how we could respect those settings inside of an extension.

**Comment by chrmarti** [comment]
@matthewjamesadam Are you changing the system proxy setting while VS Code is open? Does restarting VS Code fix the issue? We cache the proxy (https://github.com/microsoft/vscode-proxy-agent/blob/49c0f39c327ce408ce69247a3db10a2b313d44a2/src/index.ts#L69) for each URL to avoid going all the way to the main process to look up the proxy each time. (We are using an Electron API to do the lookup.)

@DavyLandman We load the root certificates from the OS (https://github.com/microsoft/vscode-proxy-agent/blob/49c0f39c327ce408ce69247a3db10a2b313d44a2/src/index.ts#L401).

**Comment by DavyLandman** [comment]
@chrmarti does this loading of the root certificates also happen for the extension host? Since we've been running into that. Although we are using `node-fetch` instead of plain `http` calls. I'll investigate further, and open an issue if needed.

**Comment by chrmarti** [comment]
@DavyLandman Yes, we indeed patch the http and https built-in Node modules in the extension host. `node-fetch` probably uses these. Separate issue appreciated if this doesn't work as expected. Thanks.

**Comment by DavyLandman** [comment]
@chrmarti just to report back, indeed, that works, we'll have to dive deeper on why some users with corporate firewalls are having issues, but this just works as intended. sorry @matthewjamesadam to high-jack the issue.

**Comment by matthewjamesadam** [comment]
> @matthewjamesadam Are you changing the system proxy setting while VS Code is open? Does restarting VS Code fix the issue? We cache the proxy (https://github.com/microsoft/vscode-proxy-agent/blob/49c0f39c327ce408ce69247a3db10a2b313d44a2/src/index.ts#L69) for each URL to avoid going all the way to the main process to look up the proxy each time. (We are using an Electron API to do the lookup.)

@chrmarti in effect, yes, the system proxy settings are changing while VSCode is open.  I am not manually changing the settings myself, but many tools (VPNs, network monitoring tools like Proxyman) work by changing the system proxy settings.  For instance, opening Proxyman changes the system proxy settings to redirect network traffic to itself (acting as a MITM effectively), while closing it reverts the proxy settings.  I use OpenVPN regularly, and whenever I enable or disable the VPN (which I do somewhat often), under the hood it is changing my system proxy settings.

Whenever I take any of these actions while VSCode is open, all extensions that rely on network resources effectively stop working.  Caching the proxy is probably what is causing this.  I am not sure if there is a way VSCode can detect a change to the system proxy settings and clear the proxy cache.  I would be happy to look into this if you gave me a few pointers.

**Comment by matthewjamesadam** [comment]
Just to give a little more context: this can happen even without explicit user actions.  OpenVPN will sometimes change its proxy settings when my laptop is closed or opened -- the end result is that sometimes when I open my laptop in the morning, VSCode's extensions are no longer working.
